### ISSUE TITLE
Document / adjust operator precedence lint

### ISSUE THREAD

**Issue by dhardy** [rust-lang/rust-clippy#13966]
### Description

The [precedence lint](https://rust-lang.github.io/rust-clippy/master/index.html#precedence) flags some stuff which is (IMO) unintuitive:

    `1 << 2 + 3`

But also some stuff which is (again, IMO) clearer:

    `u64::from(data[1]) << 32 | u64::from(data[0])`

The lint does not flag some stuff which is widely understood:

    `2 * 3 - 1`

---

The lint documentation could be more precise about which operator combinations get flagged, with a [link to the reference on operator expression precedence](https://doc.rust-lang.org/reference/expressions.html?highlight=precedence#expression-precedence).

Further, the lint could be disagglomerated or adjusted to allow the bit-shift with binary-and/or/xor case above.

### Version

```text

```

**Comment by dhardy** [comment]
Also, *is* this lint deprecated, while still being reported?
```
$ cargo clippy --all-targets -- -A precedence
    Checking rand_core v0.9.0-beta.1 (/home/dhardy/projects/rand/rand/rand_core)
warning: lint name `precedence` is deprecated and may not have an effect in the future
  |
  = help: change it to clippy::precedence
  = note: requested on the command line with `-A precedence`
  = note: `#[warn(renamed_and_removed_lints)]` on by default

warning: operator precedence can trip the unwary
   --> rand_core/src/block.rs:200:13
    |
200 |             u64::from(data[1]) << 32 | u64::from(data[0])
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider parenthesizing your expression: `(u64::from(data[1]) << 32) | u64::from(data[0])`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#precedence
    = note: `#[warn(clippy::precedence)]` on by default
```

**Comment by masklinn** [comment]
> Also, _is_ this lint deprecated, while still being reported?

The lint is not being deprecated, "un-namespaced" lints are. This is to avoid conflicts and allow different linting tools to reliably ignore one another's settings.
